Added --skip-benchmarks command-line option.

danifeist opened the pull request:
This PR adds the feature requested by https://github.com/catchorg/Catch2/issues/2392 "Command line option to skip benchmarks" by introducing the command-line option "--skip-benchmarks" so that Catch runs no benchmarks when this option is set. I'd be glad about feedback.

danifeist committed:
Added --skip-benchmarks command-line option.

horenmar commented on docs/command-line.md:
```diff
@@ -370,6 +372,12 @@ See [The LibIdentify repo for more information and examples](https://github.com/
+<a id="skip-benchmarks"></a>
+## Skip all benchmarks
+<pre>--skip-benchmarks</pre>
+
```
This should have a version placeholder (see contributing.md for how it looks).

horenmar commented on docs/command-line.md:
```diff
@@ -370,6 +372,12 @@ See [The LibIdentify repo for more information and examples](https://github.com/
+## Skip all benchmarks
+<pre>--skip-benchmarks</pre>
+
+When this option is set, all benchmarks of each and every Catch test case will be skipped so that it appears as if there were no benchmark defined.
```
How about

```suggestion
This flag tells Catch2 to skip running all benchmarks. Benchmarks in this
case mean code blocks in `BENCHMARK` and `BENCHMARK_ADVANCED` macros, not
test cases with the `[!benchmark]` tag.
```

I think this is clearer. (Also new docs use Catch2 and line length of 80, to play nicer with diffs)

horenmar commented on src/catch2/benchmark/catch_benchmark.hpp:
```diff
@@ -96,7 +96,10 @@ namespace Catch {
+                auto const* cfg = getCurrentContext().getConfig();
+                if (!cfg->skipBenchmarks()) {
+                    run();
+                }
```
I think this should happen before we create the function wrapper.

danifeist replied:
you'd place `fun = Detail::BenchmarkFunction(func);` inside the if-clause, did i got that right?

horenmar replied:
Yeah. There is no reason to make the allocation if it won't be used afterwards.

horenmar reviewed (changes requested):
This also needs tests.

It should be enough to pick some TEST_CASE that's already in the SelfTest project and has benchmarks, and then add new CTest test that invokes `SelfTest` with `--skip-benchmarks` and test spec for that one test. The test then checks that the benchmark was skipped by looking for the benchmark output.

Or add a new TEST_CASE just for this, that way is more resilient to later changes.

danifeist committed:
Merge branch 'devel' into command_line_option_skip_benchmarks

danifeist committed:
Improved documentation of --skip-benchmarks cmd line option
- Added version placeholder in documentation
- Improved description of as suggested

danifeist committed:
Added new CTest test for --skip-benchmarks cmd line option
The new CTest runs a dedicated TEST_CASE with the --skip-benchmarks option.
The TEST_CASE can only succeed when the two benchmarks inside are skipped.

danifeist committed:
Removed unused includes

danifeist committed:
Set lambda to be used if and only if benchmark is not skipped

horenmar committed:
Tighten up --skip-benchmarks test

horenmar commented on tests/CMakeLists.txt:
```diff
@@ -359,6 +359,18 @@ add_test(NAME CheckConvenienceHeaders
+  PROPERTIES
+    PASS_REGULAR_EXPRESSION "All tests passed \\(2 assertions in 1 test case\\)"
+    FAIL_REGULAR_EXPRESSION "benchmark name"
+)
```
I tightened up the tests a bit. This way the tests fail if either the assertions disappear for some reason, or if the "benchmark begins" event gets sent (even if it isn't executed afterwards).

horenmar commented:
Thanks.
